core/src/main/java/org/apache/iceberg/variants/Variants.java

```diff
@@ -0,0 +1,274 @@
+    }
+
+    LogicalType toLogicalType() {
+      return logicalType;
+    }
+
+    public Class<?> javaClass() {
+      return javaClass;
+    }
+
+    public static PhysicalType from(int primitiveType) {
+      switch (primitiveType) {
+        case Primitives.TYPE_NULL:
+          return NULL;
+        case Primitives.TYPE_TRUE:
+          return BOOLEAN_TRUE;
+        case Primitives.TYPE_FALSE:
+          return BOOLEAN_FALSE;
+        case Primitives.TYPE_INT8:
+          return INT8;
+        case Primitives.TYPE_INT16:
+          return INT16;
+        case Primitives.TYPE_INT32:
+          return INT32;
+        case Primitives.TYPE_INT64:
+          return INT64;
+        case Primitives.TYPE_DATE:
+          return DATE;
+        case Primitives.TYPE_TIMESTAMPTZ:
+          return TIMESTAMPTZ;
+        case Primitives.TYPE_TIMESTAMPNTZ:
+          return TIMESTAMPNTZ;
+        case Primitives.TYPE_FLOAT:
+          return FLOAT;
+        case Primitives.TYPE_DOUBLE:
+          return DOUBLE;
+        case Primitives.TYPE_DECIMAL4:
+          return DECIMAL4;
+        case Primitives.TYPE_DECIMAL8:
+          return DECIMAL8;
+        case Primitives.TYPE_DECIMAL16:
+          return DECIMAL16;
+        case Primitives.TYPE_BINARY:
+          return BINARY;
+        case Primitives.TYPE_STRING:
+          return STRING;
+      }
+
+      throw new UnsupportedOperationException("Unknown primitive physical type: " + primitiveType);
+    }
+  }
+
+  interface Serialized {
+    ByteBuffer buffer();
+  }
+
+  abstract static class SerializedValue implements VariantValue, Serialized {
+    @Override
+    public int sizeInBytes() {
+      return buffer().remaining();
+    }
+
+    @Override
+    public int writeTo(ByteBuffer buffer, int offset) {
+      ByteBuffer value = buffer();
+      VariantUtil.writeBufferAbsolute(buffer, offset, value);
+      return value.remaining();
+    }
+  }
+
+  static class Primitives {
+    static final int TYPE_NULL = 0;
+    static final int TYPE_TRUE = 1;
+    static final int TYPE_FALSE = 2;
+    static final int TYPE_INT8 = 3;
+    static final int TYPE_INT16 = 4;
+    static final int TYPE_INT32 = 5;
+    static final int TYPE_INT64 = 6;
+    static final int TYPE_DOUBLE = 7;
+    static final int TYPE_DECIMAL4 = 8;
+    static final int TYPE_DECIMAL8 = 9;
+    static final int TYPE_DECIMAL16 = 10;
+    static final int TYPE_DATE = 11;
+    static final int TYPE_TIMESTAMPTZ = 12; // equivalent to timestamptz
+    static final int TYPE_TIMESTAMPNTZ = 13; // equivalent to timestamp
+    static final int TYPE_FLOAT = 14;
+    static final int TYPE_BINARY = 15;
+    static final int TYPE_STRING = 16;
+
+    static final int PRIMITIVE_TYPE_SHIFT = 2;
+
+    private Primitives() {}
+  }
+
+  static final int HEADER_SIZE = 1;
+
+  enum BasicType {
+    PRIMITIVE,
+    SHORT_STRING,
+    OBJECT,
+    ARRAY
+  }
+
+  public static VariantValue from(ByteBuffer metadata, ByteBuffer value) {
```

Comment:
I think it would have to be on either `VariantValue` (which could be any value), here in `Variants`, or in `Variant`. I put it here to match the conventions from other parts of the library. For example, there are similar factory methods in many places: `Literals.from`, `Types.fromPrimitiveString`, and `Expressions.*`